[Bug 1204515] Noisy email

mythmon opened the pull request:
1. Add a lot of logging to sending email at the lowest level we can.
2. Change our logging to expose that information (I've talked with webops and we will be able to get these logs).
3. Revert the last attempt at this, since it didn't help.

r?

willkg commented on kitsune/lib/email.py:
```diff
@@ -0,0 +1,66 @@
+RealBackend = import_string(settings.EMAIL_LOGGING_REAL_BACKEND)
+
+
+class LoggingEmailBackend(RealBackend):
```
Why do the `RealBackend` thing and pull it from settings?

willkg replied:
Nevermind--I figured it out. It's because the dev backend is different than what we have in production. That probably means that to land this, we need to change the `settings_local.py` on the servers, right?

mythmon replied:
Exactly. I filed [bug 1223471](https://bugzilla.mozilla.org/show_bug.cgi?id=1223471) for the `settings_local.py` change.

willkg commented on kitsune/lib/email.py:
```diff
@@ -0,0 +1,66 @@
+            assert (
+                new_conn is None,
+                'Unexpected return from email backend open method: %r.' % (new_conn, )
+            )
```
This seems odd. Why not just handle the `None` case in an `elif` that also checks to see if the backend is the console?

mythmon replied:
If possible, I'd like to avoid explicitly checking what the backend is. Like the other assertion, this is about validating expectations. If this assert fails, then the backend didn't act like I expected it to, so things should blow up.

I tried to use assertions to validate my assumptions, and non-assertion code to deal with what I expected the backends to do.

willkg replied:
This code is mixing two different ways of checking things and there aren't any comments so I find it confusing and think it just looks awkward and possibly wrong.

It'd be easier if there were comments about what's going on and the context of the checks.

willkg commented on kitsune/lib/email.py:
```diff
@@ -0,0 +1,66 @@
+                'to_send': num_to_send,
+            })
+
+        return num_sent
```
This logs a bunch of stuff over a period of time. You might want to create a batch id and prepend that to all the lines so it's easier to look at logs and match entire batches.

mythmon replied:
That's a good idea. Do you have any clever ways of generating batch ids? My first instinct is to use a random number, or possibly something based on the when the function started.

willkg replied:
The only thing you need out of it is that you can connect logging statements that happened in the same run. I don't think it needs anything more than that. Using a timestamp from when it started or something derived from that is probably fine.

willkg commented on kitsune/lib/email.py:
```diff
@@ -0,0 +1,66 @@
+            log.error('Failed to send all emails. Sent %(sent)s out of %(to_send)s' % {
+                'sent': num_sent,
+                'to_send': num_to_send,
+            })
```
Why handle the `<=` case with an assert, but the `==` and `>` case with this `if` block?

mythmon replied:
The assert line is sanity checking. The two branches in the if statement are normal operation, where the functions are doing what they are expected to.

The assert statement is an exceptional case where the backend sent more emails that it was asked to. This is something that should never ever happen. The assertion is there to make it blow up very loudly if that supposedly-impossible case ever happens.

willkg replied:
I'm not wild about mixing two different ways of checking things. You could easily add an `elif` and throw an exception.

Comments would be good here.

willkg commented:
What's the backend we use on the servers? How can we test this code on all the relevant backends?

mythmon commented:
On the server we use the Django standard SMTP backend: `django.core.mail.backends.console.EmailBackend`. We could test that by setting up SMTP on a dev instance. I think the docs have instructions for how to do that with Gmail.

Alternatively, we could test that by putting this on stage and trying to make some emails happen.

willkg commented:
This doesn't add or augment any tests. Is the expectation that we should test this by hand before landing? If so, can you walk me through the STR you used to test?

mythmon commented:
My thinking while writing this is that the goal is to transparently proxy the email backend, so if the tests pass then the transparent proxy worked. However, I realize now that tests use a test-specific email backend, so none of the tests will exercise this code. Further, I don't know how to test logging output. I guess I could manually instantiate the class and use mock to assert it made calls to logs.

Do you think writing tests to test this in isolation is valuable? Or will the manual testing be enough?

To test this by hand, I used the site to do something that would send an email (I used private messages). I watched the console where `runserver` was running to make sure that both the normal email logging and the extra verbose email logging worked. I didn't test the SMTP version of this, but the process would be similar.

willkg commented:
nose has a logcapture thing if that helps:

http://nose.readthedocs.org/en/latest/plugins/logcapture.html

Would it help to create a django command that sends a test email? That seems like it's an easy thing to implement and it'd probably help us debug additional email-sending infrastructure.

It's definitely worth testing the smtp side of this before merging this since that's the side that runs on the servers and is the thing you want to make sure works. Maybe push this branch to -stage and test it there?

mythmon commented:
I don't think logcapture actually makes logs available in tests. I think it only deals with printing them out when they fail. I could adapt the code from logcapture though, and add another root log handler to look at during a test.

I'll take another pass at this to try and make it easier to understand and try to add some tests.

One wrinkle about testing smtp on stage is that we don't have easy access to the logs on the servers. We'd have to get webops to help with this. We'll have to do that eventually anyways, so it isn't that big a deal.

willkg commented:
My bad. testfixtures can capture logging:

https://www.pythonhosted.org/testfixtures/logging.html

The thing that does it in testfixtures is also called logcapture hence my mistake.

I think I'd want a few things:

1. (optional) batch id to make reading logs easier
2. some comments around the asserts to clarify the context and why asserts vs. ifs
3. testing the SMTP backend on -stage which verifies that this does what we need it to do in the server environments
4. some promise that this is a temporary change and not a permanent one; if it's permanent, then we really want some tests

After that, I'm game for this.

mythmon commented:
I added the batch id idea, added tests, and did some refactoring to help with that.

The biggest change is that this no longer subclasses the real backend. Instead it just proxies to an instance of it. I couldn't get it so that I could easily swap out the backends in tests when it was subclassing.

I also removed the assert statements in favor of `raise AssertionError` when something is weird. I hope this reads better.

The tests are pretty gnarly. I hope they read clearly enough, let me know if any part of them need improvements. They could probably be factored to be fewer lines of code, but I suspect they would be a bit harder to follow.

mythmon commented:
Oh, and I do expect this to be pretty short lived. Once the problem goes away I think we can nix this. I would still like to have passable test coverage for it though.

willkg commented on requirements/dev.txt:
```diff
@@ -66,5 +67,8 @@ sphinxcontrib-httpdomain==1.4.0
 tabulate==0.7.5
 
+# sha256: lM7ehUx4i_w1KArTvkn85vNiWTgp_YuOPKwe0RL-yFc
+testfixtures==4.5.0
```
I think this suggests you added tests, but I don't see any tests in this PR.

mythmon replied:
Oops.

mythmon commented:
Added in the missing file with tests ^

willkg commented on kitsune/lib/tests/test_email.py:
```diff
@@ -0,0 +1,227 @@
+
+    def send_messages(self, messages):
+        super(FlakyMockEmailBackend, self).send_messages(messages)
+        return len(messages) / 2
```
Interesting!

willkg commented:
I think this looks good. Definitely worth testing with the SMTP backend on -stage and making sure it works there. r+

mythmon commented:
I just realized: landing this code doesn't effect anything on stage since we haven't updated `settings_local.py` on stage. I'm going to merge this, and deploy it to stage and prod. We can change the setting on stage first and verify again.
